src/git/mem.ml

```diff
@@ -0,0 +1,273 @@
+(*
+ * Copyright (c) 2013-2017 Thomas Gazagnaire <thomas@gazagnaire.org>
+ * and Romain Calascibetta <romain.calascibetta@gmail.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Lwt.Infix
+
+let err_not_found n k =
+  let str = Printf.sprintf "Git.Mem.%s: %s not found" n k in
+  Lwt.fail (Invalid_argument str)
+
+module type S =
+sig
+  module Hash
+    : Ihash.S
+
+  module Path
+    : Path.S
+
+  module Inflate
+    : S.INFLATE
+
+  module Deflate
+    : S.DEFLATE
+
+  module Buffer
+    : Value.BUFFER
+
+  module Value
+    : Value.RAW
+      with module Hash = Hash
+       and module Inflate = Inflate
+       and module Deflate = Deflate
+       and module Buffer = Buffer
+
+  module Reference
+    : Reference.S
+      with module Hash = Hash
+       and module Path = Path
+
+  type t
+
+  val create : ?root:string -> ?dot_git:string -> ?level:int -> unit -> t Lwt.t
+
+  val dot_git : t -> string
+  val root : t -> string
+  val level : t -> int
+
+  val dump : Format.formatter -> t -> unit Lwt.t
+  val contents : t -> (Hash.t * Value.t) list Lwt.t
+  val size : t -> Hash.t -> int option Lwt.t
+  val read : t -> Hash.t -> Value.t option Lwt.t
+  val read_exn : t -> Hash.t -> Value.t Lwt.t
+  val mem : t -> Hash.t -> bool Lwt.t
+  val list : t -> Hash.t list Lwt.t
+  val write : t -> Value.t -> Hash.t Lwt.t
+  val references : t -> Reference.t list Lwt.t
+  val mem_reference : t -> Reference.t -> bool Lwt.t
+  val read_reference : t -> Reference.t -> Hash.t option Lwt.t
+  val read_reference_exn : t -> Reference.t -> Hash.t Lwt.t
+  val write_head : t -> Reference.head_contents -> unit Lwt.t
+  val read_head : t -> Reference.head_contents option Lwt.t
+  val write_reference : t -> Reference.t -> Hash.t -> unit Lwt.t
+  val remove_reference : t -> Reference.t -> unit Lwt.t
+  val test_and_set_reference : t -> Reference.t -> test:Hash.t option -> set:Hash.t option -> bool Lwt.t
+  val read_inflated : t -> Hash.t -> string option Lwt.t
+  val write_inflated : t -> Cstruct.t -> Hash.t Lwt.t
```

Comment:
Indeed, some parts of the `mem.ml` don't need to be wrapped in a `Lwt.t`. But we need to keep in our mind the compatibility between `store.ml` (which `Lwt.t` is needed - because we interact with the file-system) and `mem.ml` (and share the same API). So it's the global point (and this signature is come from the previous version of [ocaml-git](https://github.com/mirage/ocaml-git/blob/master/src/git/store.ml#L46)).

However, for this specific function, in the current version of the implementation of the `store.ml` with a file-system, I don't provide a `write_inflate` but only a [`write`](https://github.com/mirage/ocaml-git/blob/master/src/git/store.ml#L30) function.

The better question should be: Is it necessary to expose this function ? And, if the response is yes, how we can expose this kind of computation ? In fact, in my implementation, the signature should be:

```ocaml
val write_inflated_s : t -> Cstruct.t -> (Hash.t * int, error) result Lwt.t
```

`_s` notices than we use the _state-defined_ buffer (instead a `write_inflate_p : t -> Cstruct.t -> ztmp:Cstruct.t -> raw:Cstruct.t -> Cstruct.t -> (Hash.t * int, error) result Lwt.t`), the `Cstruct.t` is the already-serialized Git object, and we will return the `Hash.t` produced and how many byte(s) we wrote.

Finally, I think it's not needed (for the internal purpose of `ocaml-git`) to have this function (but a `read_inflated` function is needed in several points of the PACK encoder - so necessary in both-side, `store.ml` and `mem.ml`). Then, may be some peoples use `ocaml-git` and this function. We need to take care about that.